File: Java/4kyu/Human Readable Duration Format/solution.java
import java.util.Map;
import java.util.LinkedHashMap;
import java.lang.StringBuilder;

public class TimeFormatter {
    
    public static String formatDuration(int seconds) {
      System.out.println(seconds);
      LinkedHashMap<String, Integer> results = new LinkedHashMap<String, Integer>();
      String replacement = "";
      StringBuilder timeResult = new StringBuilder();
  
      Map<String, Integer> times = Map.of(
        "years", 365 * 24 * 60 * 60,
        "days", 24 * 60 * 60,
        "hours", 60* 60,
        "minutes", 60,
        "seconds", 1
      );    
  
      if(seconds == 0)
        return "now";
  
      results.put("year", seconds / times.get("years"));
      results.put("day", (seconds - times.get("years")*results.get("year"))/times.get("days"));
      results.put("hour", (seconds - times.get("years")*results.get("year") - times.get("days")*results.get("day"))/times.get("hours"));
      results.put("minute", (seconds - times.get("years")*results.get("year") - times.get("days")*results.get("day") - times.get("hours")*results.get("hour")) / times.get("minutes"));
      results.put("second", seconds - times.get("years")*results.get("year") - times.get("days")*results.get("day") - times.get("hours")*results.get("hour") - times.get("minutes")*results.get("minute"));
  
      // Removing times with value 0
      results.values().removeIf(f -> f == 0);
  
      // Checks if the date should be plural
      results.forEach((k, v) -> {
        if(v > 1)
          k += "s";
        timeResult.append(v + " " + k + ", ");
      });
  
      // Replaces the two last ', ' ocurrences with '' and ' and ', respectively
      for(int i = 0; i < 2; i++){
        int lastOcurrence = timeResult.lastIndexOf(", ");
  
        if(lastOcurrence != -1) {
          if(i != 0)
            replacement = " and ";
  
          timeResult.replace(lastOcurrence, lastOcurrence+2, replacement);
        }
      }
  
      return timeResult.toString();
    }
}

File: PHP/4kyu/Human Readable Duration Format/solution.php
<?php
function format_duration($seconds) {
  $times = array("year" => 365 * 24 * 60 * 60, 
                 "day" => 24 * 60 * 60,
                 "hour"=> 60 * 60,
                 "minute"=> 60,
                 "second" => 1);

    if(empty($seconds))
        return "now";

  $years = intdiv($seconds, $times["year"]);
  $days = intdiv(($seconds - $times["year"]*$years), $times["day"]);
  $hours = intdiv(($seconds - $times["year"]*$years - $times["day"]*$days), $times["hour"]);
  $minutes = intdiv(($seconds - $times["year"]*$years - $times["day"]*$days - $times["hour"]*$hours), $times["minute"]);
  $seconds = $seconds - $times["year"]*$years - $times["day"]*$days - $times["hour"]*$hours - $times["minute"]*$minutes;
  
  $chunks = array("year" => $years, "day" => $days, "hour" => $hours, "minute" => $minutes, "second" => $seconds);
  
  $firstiteration = 0;
  $string = "";
  foreach($chunks as $key => $time) {
    $date = $key;
    if($time) {
      if($firstiteration)
        $string .= ", ";

      if($time > 1)
        $date = $key . "s";

      $string .= strval($time) . " " . $date;
      $firstiteration = 1;
    }
  }

  $array = explode(", ", $string);
  return join(' and ', array_filter(array_merge(array(join(', ', array_slice($array, 0, -1))), array_slice($array, -1)), 'strlen'));
}
?>